src/Raven.Server/Indexing/VoronIndexInput.cs

```diff
@@ -10,6 +11,221 @@
+        }
+
+        public override string ToString()
+        {
+            return _name;
+        }
+
+        public override void SetBufferSize(int newSize)
+        {
+            if (buffer == null)
+                return;
+            
+            if (newSize <= buffer.Length)
+                return;
+
+            var newBuffer = ArrayPool<byte>.Shared.Rent(newSize);
+            Array.Copy(buffer, newBuffer, buffer.Length);
+            var oldBuffer = buffer;
+            buffer = newBuffer;
+            ArrayPool<byte>.Shared.Return(oldBuffer);
+        }
+
+        private void OpenInternal(Transaction transaction)
+        {
+            if (transaction.IsWriteTransaction == false)
+            {
+                if (transaction.LowLevelTransaction.ImmutableExternalState is IndexTransactionCache cache)
+                {
+                    if (cache.DirectoriesByName.TryGetValue(_directory.Name, out var files))
+                    {
+                        if (files.ChunksByName.TryGetValue(_name, out var details))
+                        {
+                            // we don't dispose here explicitly, the fileName needs to be
+                            // alive as long as the transaction is
+                            Slice.From(transaction.Allocator, _name, out Slice fileName);
+                            _stream = new VoronStream(fileName, details, transaction.LowLevelTransaction);
+                            return;
+                        }
+                    }
+                }
+            }
+
+            var fileTree = transaction.ReadTree(_tree);
+            if (fileTree == null)
+                throw new FileNotFoundException($"Could not find '{_tree}' tree for index input", _name);
+
+            using (Slice.From(transaction.Allocator, _name, out Slice fileName))
+            {
+                _stream = fileTree.ReadStream(fileName);
+                if (_stream == null)
+                    throw new FileNotFoundException("Could not find index input", _name);
+            }
+        }
+
+        public override object Clone(IState s)
+        {
+            var state = s as VoronState;
+            if (state == null)
+            {
+                ThrowStateNullException();
+                return null;
+            }
+
+            ThrowIfDisposed(state);
+
+            var clone = (VoronBufferedInput)base.Clone(s);
+            GC.SuppressFinalize(clone);
+            clone._isOriginal = false;
+
+            clone.OpenInternal(state.Transaction);
+            clone._stream.Position = _stream.Position;
+
+            return clone;
+        }
+        
+        protected override void Dispose(bool disposing)
+        {
+            if (_isDisposed)
+                return;
+
+            GC.SuppressFinalize(this);
+            
+            if (buffer != null)
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
If I'm not the original, I will return the buffer twice, no?

Reply:
No, every copy has its own buffer by design.